### PR TITLE
Provide pathId param on browser.js

### LINKED ISSUE #29310: Requested default folder is not selected on opening media gallery (current_tree_path parameter)
### Preconditions (*)
1. Install Magento 2.4.x

### Steps to reproduce (*)
1. Open media gallery from the category image field
2. Open media gallery from the WYSIWYG

### Expected result (*)
If no filters were applied (except directory) the default directory should be selected based on `current_tree_path` parameter
1. catalog/category directory is selected by default
2. WYSIWYG directory is selected by default

### Actual result (*)
The previously selected directory is selected by default

### Additional details
The directory that should be selected by default is provided via `current_tree_path` parameter

### PR DESCRIPTION
**Pull request by Nazar65**
### Fixed Issues (if relevant)

1. Fixes magento/adobe-stock-integration#1335: Requested default folder is not selected on opening media gallery (current_tree_path parameter)
### Manual testing scenarios (*)
1. Open media gallery from the category image field
2. Open media gallery from the WYSIWYG

### Expected result (*)
1. catalog/category directory is selected by default
2. WYSIWYG directory is selected by default
### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

File: app/code/Magento/Cms/Model/Wysiwyg/Gallery/DefaultConfigProvider.php [hunk 1]
```diff
@@ -78,6 +78,8 @@ public function getConfig(\Magento\Framework\DataObject $config) : \Magento\Fram
         if (is_string($this->currentTreePath)) {
             $fileBrowserUrlParams = [
                 'current_tree_path' => $this->imagesHelper->idEncode($this->currentTreePath),
+                'tree_path' => $this->currentTreePath . '&'
```
Comment (gabrieldagama) [inline app/code/Magento/Cms/Model/Wysiwyg/Gallery/DefaultConfigProvider.php]: Can't we just decode the value of currentTreePath on the frontend instead of adding the information here?

File: lib/web/mage/adminhtml/browser.js [hunk 2]
```diff
@@ -96,6 +98,8 @@ define([
          * @param {String} url
          */
         setTargetElementId: function (options, url) {
+            this.currentTreePath = url.match(/\/tree_path\/([\s\S].*?)(&|%26)/)[1];
```
Comment (gabrieldagama) [inline lib/web/mage/adminhtml/browser.js]: This function is a setter, it should only set the targetElementId property, please create setters for your custom properties if required.

File: app/code/Magento/Cms/Test/Mftf/Test/AdminVerifyDefaultCurrentTreePathTest.xml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  * See COPYING.txt for license details.
+  */
+-->
+<tests xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/testSchema.xsd">
+    <test name="AdminVerifyDefaultCurrentTreePathTest">
+        <annotations>
+            <features value="Cms"/>
+            <group value="Cms"/>
+            <stories value="MediaGallery Verify Preselcted folders on Category Page"/>
```
Comment (gabrieldagama) [inline app/code/Magento/Cms/Test/Mftf/Test/AdminVerifyDefaultCurrentTreePathTest.xml]:
```suggestion
            <stories value="MediaGallery verify preselected folders on Category Page"/>
```

File: app/code/Magento/Cms/Test/Mftf/Test/AdminVerifyDefaultCurrentTreePathTest.xml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+            <title value="Admin should be see preselcted default folder in Media Gallery"/>
```
Comment (gabrieldagama) [inline app/code/Magento/Cms/Test/Mftf/Test/AdminVerifyDefaultCurrentTreePathTest.xml]:
```suggestion
            <title value="Admin should be able to see preselected default folder in Media Gallery"/>
```

File: app/code/Magento/Cms/Test/Mftf/Test/AdminVerifyDefaultCurrentTreePathTest.xml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+            <description value="Admin should be see preselcted default folder in Media Gallery"/>
```
Comment (gabrieldagama) [inline app/code/Magento/Cms/Test/Mftf/Test/AdminVerifyDefaultCurrentTreePathTest.xml]:
```suggestion
            <description value="Admin should be able to see preselected default folder in Media Gallery"/>
```

File: app/code/Magento/Cms/Test/Mftf/Test/AdminVerifyDefaultCurrentTreePathTest.xml [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+        <after>
+            <actionGroup ref="AdminLogoutActionGroup" stepKey="logout"/>
+        </after>
+
+        <actionGroup ref="NavigateToMediaGalleryActionGroup" stepKey="navigateToMediaGallery"/>
+        <!-- Verify catalog/category folder selected -->
+        <waitForPageLoad stepKey="waitForFoldersLoad"/>
+        <seeElement selector="{{MediaGallerySection.higlightedFolder('Y2F0YWxvZy9jYXRlZ29yeQ--')}}" stepKey="assertCategoryFolder"/>
+        <actionGroup ref="AdminCloseMediaGalleryActionGroup" stepKey="closeMEdiaGallery"/>
```
Comment (gabrieldagama) [inline app/code/Magento/Cms/Test/Mftf/Test/AdminVerifyDefaultCurrentTreePathTest.xml]:
```suggestion
        <actionGroup ref="AdminCloseMediaGalleryActionGroup" stepKey="closeMediaGallery"/>
```

File: app/code/Magento/Ui/view/base/web/js/form/element/image-uploader.js [hunk 7]
```diff
@@ -78,6 +78,7 @@ define([
                 '/type/image/?isAjax=true';
 
             if (this.mediaGallery.initialOpenSubpath) {
+                openDialogUrl += '/tree_path/' + this.mediaGallery.initialOpenSubpath;
```
Comment (gabrieldagama) [inline app/code/Magento/Ui/view/base/web/js/form/element/image-uploader.js]: If we remove the tree_path key from the DefaultConfigProvider this won't be necessary, right?

File: app/code/Magento/Cms/Test/Mftf/Section/TinyMCESection/MediaGallerySection.xml [hunk 8]
```diff
@@ -34,6 +34,7 @@
         <element name="WysiwygArrow" type="button" selector="#d3lzaXd5Zw-- &gt; .jstree-icon"/>
         <element name="checkIfWysiwygArrowExpand" type="button" selector="//li[@id='d3lzaXd5Zw--' and contains(@class,'jstree-closed')]"/>
         <element name="confirmDelete" type="button" selector=".action-primary.action-accept"/>
+        <element name="higlightedFolder" type="block" selector="//li[@id='{{folderId}}']/a[@class='jstree-clicked']" parameterized="true"/>
```
Comment (gabrieldagama) [inline app/code/Magento/Cms/Test/Mftf/Section/TinyMCESection/MediaGallerySection.xml]:
```suggestion
        <element name="highlightedFolder" type="block" selector="//li[@id='{{folderId}}']/a[@class='jstree-clicked']" parameterized="true"/>
```

File: app/code/Magento/Cms/Test/Mftf/Test/AdminVerifyDefaultCurrentTreePathTest.xml [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        </before>
+
+        <after>
+            <actionGroup ref="AdminLogoutActionGroup" stepKey="logout"/>
+        </after>
+
+        <actionGroup ref="NavigateToMediaGalleryActionGroup" stepKey="navigateToMediaGallery"/>
+        <!-- Verify catalog/category folder selected -->
+        <waitForPageLoad stepKey="waitForFoldersLoad"/>
+        <seeElement selector="{{MediaGallerySection.higlightedFolder('Y2F0YWxvZy9jYXRlZ29yeQ--')}}" stepKey="assertCategoryFolder"/>
```
Comment (gabrieldagama) [inline app/code/Magento/Cms/Test/Mftf/Test/AdminVerifyDefaultCurrentTreePathTest.xml]:
```suggestion
        <seeElement selector="{{MediaGallerySection.highlightedFolder('Y2F0YWxvZy9jYXRlZ29yeQ--')}}" stepKey="assertCategoryFolder"/>
```

File: app/code/Magento/Cms/Test/Mftf/Test/AdminVerifyDefaultCurrentTreePathTest.xml [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+        <actionGroup ref="AdminCloseMediaGalleryActionGroup" stepKey="closeMEdiaGallery"/>
+        <actionGroup ref="AdminOpenMediaGalleryTinyMce4ActionGroup" stepKey="opneMediaGalleryFromTinyMce4"/>
+         <!-- Verify wysiwyg folder selected -->
+        <seeElement selector="{{MediaGallerySection.higlightedFolder('d3lzaXd5Zw--')}}" stepKey="assertWysiwygFolder"/>
```
Comment (gabrieldagama) [inline app/code/Magento/Cms/Test/Mftf/Test/AdminVerifyDefaultCurrentTreePathTest.xml]:
```suggestion
        <seeElement selector="{{MediaGallerySection.highlightedFolder('d3lzaXd5Zw--')}}" stepKey="assertWysiwygFolder"/>
```

**Review by gabrieldagama — CHANGES_REQUESTED** [review]
Hi @Nazar65 thanks for your pull request, it looks good! Please follow below some comments. Thanks!

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

File: app/code/Magento/Cms/view/adminhtml/web/js/folder-tree.js [hunk 11]
```diff
@@ -101,8 +104,45 @@ define([
...
+                        }
+                    });
+                    paths.unshift('root');
+                    paths.reverse();
+
+                    return paths;
                 };
 
+            $(window).on('reload.MediaGallery', function () {
+                encodedPath = Base64.decode(window.MediabrowserUtility.pathId.replace(/--|,,/, '==')).split('/');
```
Comment (gabrieldagama) [inline app/code/Magento/Cms/view/adminhtml/web/js/folder-tree.js]: I believe this variable name should be decodedPath since it is decoding the pathId, correct?

File: app/code/Magento/Cms/view/adminhtml/web/js/folder-tree.js [hunk 12]
```diff
@@ -101,8 +104,45 @@ define([
...
+                pathId =  window.MediabrowserUtility.pathId.replace(',,', '--');
+
+                tree.jstree('deselect_all');
+
+                if (encodedPath.length > 1) {
+                    path =  _parseCurrentPath(encodedPath);
```
Comment (gabrieldagama) [inline app/code/Magento/Cms/view/adminhtml/web/js/folder-tree.js]: The _parseCurrentPath doesn't accept any parameter, I believe we can remove the encodedPath from here.

**Review by gabrieldagama — COMMENTED** [review]
Hi @Nazar65, thanks for your changes! Please follow some small comments below.

File: app/code/Magento/Cms/view/adminhtml/web/js/folder-tree.js [hunk 13]
```diff
@@ -101,8 +104,45 @@ define([
...
+                _parseCurrentPath = function () {
+                    var paths = [];
+
+                    $.each(encodedPath, function (i, val) {
+                        var isLastElement = i === encodedPath.length - 1;
+
+                        if (isLastElement) {
+                            paths[i] = window.MediabrowserUtility.pathId.replace(',,', '--');
+                        } else {
+                            paths[i] = Base64.encode(val).replace('==', '--');
```
Comment (gabrieldagama) [inline app/code/Magento/Cms/view/adminhtml/web/js/folder-tree.js]: I think the replace here isn't correct and may cause other problems, please follow below the way PHP is handling the encode, I believe we have to perform the same behavior on JS.

file: app/code/Magento/Cms/Helper/Wysiwyg/Images.php
method: idEncode
line: 304
`
        return strtr(base64_encode($string), '+/=', ':_-');
`

**Comment by Nazar65** [comment]
@gabrieldagama thank you for review, could you please review new changes ?

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

File: dev/tests/js/jasmine/tests/lib/mage/browser.test.js [hunk 14]
```diff
@@ -41,8 +42,8 @@ define([
                         }
                     };
                 });
-            obj.openDialog('instance/url', 100, 100, 'title', options);
-            obj.openDialog('instance/url', 100, 100, 'title', options);
+            obj.openDialog(openUrl, 100, 100, 'title', options);
+            obj.openDialog(openUrl, 100, 100, 'title', options);
             expect($.ajax.calls.count()).toBe(1);
```
Comment (gabrieldagama) [inline dev/tests/js/jasmine/tests/lib/mage/browser.test.js:48]: Can we verify if the pathId is set on these tests?

**Review by gabrieldagama — CHANGES_REQUESTED** [review]
Hi @Nazar65, thanks for your pull request, it looks good. Please follow the comment below.

File: dev/tests/js/jasmine/tests/lib/mage/browser.test.js (hunk 14 above)
Comment (Nazar65) [reply dev/tests/js/jasmine/tests/lib/mage/browser.test.js:48]: actually this test will fail if in url have not current_tree_path, so they cover this case

Comment (Nazar65) [reply dev/tests/js/jasmine/tests/lib/mage/browser.test.js:48]: because they always set pathId from url, so this case is covered

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run WebAPI Tests

**Review by gabrieldagama — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @gabrieldagama, thank you for the review.
[ENGCOM-7573](https://jira.corp.magento.com/browse/ENGCOM-7573) has been created to process this Pull Request
:eight_spoked_asterisk: @gabrieldagama, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Bravo** [comment]
:heavy_check_mark: QA Passed

Testing Scenario
Precondition:
1. Adobe Stock Integration branch cloned and required by the composer. (You can find the installation guide  [here](https://github.com/magento/adobe-stock-integration/wiki/Installation))
2. Switch Adobe Stock to [this](https://github.com/magento/adobe-stock-integration/pull/1350) if it is not merged yet

Steps:
1. Open Magento Admin
2. Go to Stores - Configuration - Advanced tab - System and select Enabled Yes for Enhanced Media Gallery. Save Config
3. Go to Catalog - Categories
(the Default Category is selected)
4. Expand Content
5. Click Select from Gallery button near the Category Image 
Expected Result: Catalog/Category directory is selected
cat-cat_2205

6. Close Media Gallery 
7. Click Insert/edit image from Description editor
8. From the opened Insert/edit image window click Search button near the Source field
Expected Result: wysiwyg directory is selected
cat-wysiwyg_2205

**Comment by chalov-anton** [comment]
Added test steps to already existing scenario on cucumber studio https://studio.cucumber.io/projects/131313/test-plan/folders/1054245/scenarios/4523889
